xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+
+
+def _validate_region_lon_bounds(grid_lon_bounds, region_lon_bounds):
+    """Determines if values in a list of longitude coordinates falls within
+    longitudinal grid cell boundaries of a specified longitude axis.
+
+    Operations include:
+
+    - Check to determine whether longitude values are within the boundaries of
+    the target longitude axis
+
+    Parameters
+    ----------
+    grid_lon_bounds : xr.DataArray
+        longitude axis [len(lon), 2]
+    region_lon_bounds : float | int | list | array
+        list of longitude values to validate
+
+    Returns
+    -------
+    boolean
+        True if all input longitude bounds are within the boundaries of the
+        target longitudinal axis
+
+    Examples
+    --------
+    Import:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+
+    Open dataset and dataarray:
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+
+    Specify some regional bounds:
+    >>> regional_bounds = [-170, -120]
+
+    Validate:
+    >>> in_bounds = _validate_region_lon_bounds(tas.bounds.lon,
+                                                regional_bounds)
+    """
+    # ensure iterable array
+    region_lon_bounds = np.asarray(region_lon_bounds)
+    region_lon_bounds = np.atleast_1d(region_lon_bounds)
+    # get the max / min bounds for each grid cell
+    min_grid_lon_bounds = np.min(grid_lon_bounds, axis=1)
+    max_grid_lon_bounds = np.max(grid_lon_bounds, axis=1)
+    # iterate over regional bounds and check to see if they fall within
+    # any grid cell bounds.
+    in_bounds = True
+    for lon in region_lon_bounds:
+        inds = np.where((min_grid_lon_bounds <= lon) &
+                        (max_grid_lon_bounds >= lon))[0]
+        if len(inds) == 0:
+            in_bounds = False
+    return in_bounds
+
+
+def _switch_lon_bounds(lon, totype='to360'):
+    """Switch longitude axis orientation either 'to360' (0 to 360) or 'to180'
+    (0 to 180).
+
+    Parameters
+    ----------
+    lon : array
+        array of longitude values to convert
+    totype : str, optional
+        string specifying conversion type ('to180' or 'to360')
+
+    Returns
+    -------
+    array
+        Array of converted longitude values
+
+    Notes
+    -----
+    This does not re-order the values in any way; it only converts the values
+    in-place between longitude conventions (0 to 360 or -180 to 180).
+    """
+    orig_type = type(lon)  # note original datatype
+    lon = np.array(lon)  # convert to array
+    # convert to specified longitude axis orientation
+    if totype == 'to360':
+        inds = np.where(lon < 0)[0]
+        lon[inds] = lon[inds] + 360
+    elif totype == 'to180':
+        inds = np.where(lon > 180)[0]
+        lon[inds] = lon[inds] - 360
+    # switch back to original data type
+    # REVIEW: Is there an easier way to be flexible to the datatype provided?
```

Comment:
There must be a better way to do this - right?

Reply:
Here's going to 180: https://stackoverflow.com/a/53561230
xarray also provides the same formula: https://xarray.pydata.org/en/stable/generated/xarray.DataArray.assign_coords.html
```
da.coords['lon'] = (da.coords['lon'] + 180) % 360 - 180
da = da.sortby(da.lon)
```
You can do something similar for going to 360.

The benefit of using this formula is that you don't to filter using `.where()`, which might make it quicker.

Reply:
Here's a refactored version of this function.

Notes:

- Generally, it is better to work directly on the `xarray.DataArray`  instead of converting it into a `numpy.ndarray`. You maintain the attributes of the `DataArray` and don't need to perform type conversion. If xarray does not support an operation, you can still perform `numpy.ndarray` operations directly on the `xarray.DataArray` due to the interoperability.

My refactored function:
- Pass the entire xarray DataArray instead of just the `lon` coordinate. 
- The mid-point is -180, not 180 like the original `_switch_lon_bounds()`.

https://github.com/tomvothecoder/xcdat/blob/911419a8580688cbdad4a362d72e6e804b387272/xcdat/average_tv.py#L18-L77